src/main/java/com/arpnetworking/metrics/mad/parsers/PrometheusToRecordParser.java

```diff
@@ -0,0 +1,245 @@
+     * */
+    Optional<ParseResult> parseUnit(final String name) {
+        if (!_interpretUnits) {
+            return Optional.empty();
+        }
+        final StringBuilder builder = new StringBuilder();
+        for (int i = name.length() - 1; i >= 0; i--) {
+            final char ch = name.charAt(i);
+            if (ch == '_') {
+                final String key = builder.toString();
+                if (PROMETHEUS_AGGREGATION_KEYS.contains(key)) {
+                    builder.setLength(0); //reset builder
+                } else {
+                    final Unit value = UNIT_MAP.get(key);
+                    if (value != null) {
+                        final String newName = name.substring(0, i).concat(name.substring(i + 1 + key.length()));
+                        return Optional.of(new ParseResult(value, newName));
+                    } else {
+                        return Optional.empty();
+                    }
+                }
+            } else {
+                builder.append(ch);
+            }
+        }
+        final String possibleUnit = builder.toString();
+        final Unit value = UNIT_MAP.get(possibleUnit);
+        if (value != null) {
+            final String newName = name.substring(Math.min(possibleUnit.length() + 1, name.length()));
+            return Optional.of(new ParseResult(value, newName));
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    @Override
+    public List<Record> parse(final HttpRequest data) throws ParsingException {
+        final List<Record> records = Lists.newArrayList();
+        final byte[] uncompressed;
+        try {
+            uncompressed = Snappy.uncompress(data.getBody().toArray());
+        } catch (final IOException e) {
+            throw new ParsingException("Failed to decompress snappy stream", data.getBody().toArray(), e);
+        }
+        try {
+            final Remote.WriteRequest writeRequest = Remote.WriteRequest.parseFrom(uncompressed);
+            for (final TimeSeries timeSeries : writeRequest.getTimeseriesList()) {
+                Optional<String> nameOpt = Optional.empty();
+                final ImmutableMap.Builder<String, String> dimensionsBuilder = ImmutableMap.builder();
+                for (Types.Label label : timeSeries.getLabelsList()) {
+                    if ("__name__".equals(label.getName())) {
+                        final String value = label.getValue();
+                        nameOpt = Optional.ofNullable(value);
+                    } else {
+                        dimensionsBuilder.put(label.getName(), label.getValue());
+                    }
+                }
+                if (!nameOpt.isPresent()) {
+                    throw new ParsingException("Could not find the metric name", data.getBody().toArray());
+                }
+                final String metricName = nameOpt.get().trim();
+                if (metricName.isEmpty()) {
```

Comment:
You should move this code below the `parseUnit` call (to be renamed; see above) and then you can enforce this condition on the input or the parsed metric name which seems reasonable to me (in answer to your question about what to do about an empty parsed metric name).

Reply:
^ if clients don't like that, they can turn off name parsing.